compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3629,8 +3629,12 @@ class Typer extends Namer
     }
 
   private def checkStatementPurity(tree: tpd.Tree)(original: untpd.Tree, exprOwner: Symbol)(using Context): Unit =
-    if (!tree.tpe.isErroneous && !ctx.isAfterTyper && isPureExpr(tree) &&
-        !tree.tpe.isRef(defn.UnitClass) && !isSelfOrSuperConstrCall(tree))
+    if !tree.tpe.isErroneous
+      && !ctx.isAfterTyper
+      && !tree.isInstanceOf[Inlined]
+      && isPureExpr(tree)
+      && !isSelfOrSuperConstrCall(tree)
```

Comment:
`tree.isInstanceOf[Inlined]` : can an inlined tree be pure? What about handle inlined trees in `isPureExpr`?

Reply:
The Inlined trees can be pure which is correctly identified in isPureExpr. Here we special case a
Inlined calls that inserted a pure expression in a statement possition as these are only puede due to other constraints that may change. For example the assert can take a reference to an Inlined true value which converts it in a noop and we do not want to warn that this is a pure expression in statement position.

Reply:
Thanks for the explanation. It might be good to add a short comment after the line.